Operator updates to support generic math (#29794)
dotnet/docs: docs/csharp/language-reference/operators/arithmetic-operators.md

Those operators are supported by all [integral](../builtin-types/integral-numeric-types.md) and [floating-point](../builtin-types/floating-point-numeric-types.md) numeric types.

> Beginning with C# 11, you can define `checked` and unchecked variants of all these operators except the `%` operator. Generally, a `checked` operator is expected to throw an exception when the result overflows the destination type. An unchecked operator doesn't throw an exception. Instead it *truncates* the result. How the result is truncated depends on the specific type. The types in the .NET runtime follow these semantics. The compiler doesn't enforce these requirements, but developers should follow those guidelines in their types. The `checked` operator includes the `checked` keyword. The unchecked operator doesn't include that modifier, as the following example shows:

Reviewer: Unary plus also cannot be `checked`

> Beginning with C# 11, you can define `checked` and unchecked variants of all these operators except the `%` operator. Generally, a `checked` operator is expected to throw an exception when the result overflows the destination type. An unchecked operator doesn't throw an exception. Instead it *truncates* the result. How the result is truncated depends on the specific type. The types in the .NET runtime follow these semantics. The compiler doesn't enforce these requirements, but developers should follow those guidelines in their types. The `checked` operator includes the `checked` keyword. The unchecked operator doesn't include that modifier, as the following example shows:

Reviewer: I wouldn't use the word `truncate`. That typically has a well defined meaning, often meaning to wrap around. Where-as saturate is the terminology commonly used for clamping (that is fixing to the minimum or maximum value instead).

I would probably describe this as how a type overflows is undefined. Some types, like `int` will truncate and some types like `float` will saturate instead. Other types, like `decimal`, will always throw on overflow and do not respect checked or unchecked contexts.

> :::code language="csharp" source="snippets/shared/ArithmeticOperators.cs" id="CheckedOperator":::

Reviewer 2: Is this required? I would assume that if you put the `checked` keyword in the operator, you wouldn't need to do it in the code below, because it's only called in the checked context. I've not read the spec, so I'm just pointing out what doesn't make sense.

![image](https://user-images.githubusercontent.com/67293991/173694151-c93bdee2-2efc-459e-a199-a3318711f375.png)

Author: Yes. That guarantees that the `operator +` for the `int` type executes in a checked context.

> If you provide the unchecked operator, that overloaded operator is called in both a `checked` and `unchecked` context. You must provide an unchecked operator if you provide a `checked` operator.

Reviewer: Should this read
> If you only provide the unchecked operator?

I'm also not sure if we should describe the `static operator +` as `unchecked` or not. When its the only operator defined, the closest one might describe it is that it's oblivious. It has a behavior and that behavior is used everywhere. That might be to throw on overflow (such as `decimal`) and it might be to do something else (like truncate or saturate).

When a `checked` operator is defined, then the implication is that the other operator is now the "unchecked" variant, that is it will be called in an unchecked context.

Reviewer: Perhaps we should then distinguish that when you only provide one operator, `static operator +`, that it is just the operator.

But when you are referring to the pair (particularly since defining `checked` requires you provide both) they are `checked` and `unchecked`.

Author: I rewrote that entire paragraph.

> Those operators (except the `++` and `--` operators) are defined for the `int`, `uint`, `long`, and `ulong` types. When operands are of other integral types (`sbyte`, `byte`, `short`, `ushort`, or `char`), their values are converted to the `int` type, which is also the result type of an operation. When operands are of different integral or floating-point types, their values are converted to the closest containing type, if such a type exists. For more information, see the [Numeric promotions](~/_csharpstandard/standard/expressions.md#1147-numeric-promotions) section of the [C# language specification](~/_csharpstandard/standard/README.md). The `++` and `--` operators are defined for all integral and floating-point numeric types and the [char](../builtin-types/char.md) type.

Reviewer: > except the `++` and `--` operators

Not sure I understand that bit. `++` and `--` work for `int`, `uint`, `long`, and `ulong` (as well as all the other primitives).

It is the case that for `++`, `--`, and all the `+=` and similar assignment operators that they don't upcast the small types (that is `byte += byte` and `byte++` stays `byte`, it doesn't become `int`).

> Those operators are defined for the `int`, `uint`, `long`, and `ulong` types. When operands are of other integral types (`sbyte`, `byte`, `short`, `ushort`, or `char`), their values are converted to the `int` type, which is also the result type of an operation. When operands are of different integral or floating-point types, their values are converted to the closest containing type, if such a type exists. For more information, see the [Numeric promotions](~/_csharpstandard/standard/expressions.md#1147-numeric-promotions) section of the [C# language specification](~/_csharpstandard/standard/README.md). The `++` and `--` operators are defined for all integral and floating-point numeric types and the [char](../builtin-types/char.md) type. The `++`, `--` and all compound arithmetic operators (such as `+=`) do not convert their arguments to `int` or have the result type as `int`.

Reviewer 2: ```suggestion
Those operators are defined for the `int`, `uint`, `long`, and `ulong` types. When operands are of other integral types (`sbyte`, `byte`, `short`, `ushort`, or `char`), their values are converted to the `int` type, which is also the result type of an operation. When operands are of different integral or floating-point types, their values are converted to the closest containing type, if such a type exists. For more information, see the [Numeric promotions](~/_csharpstandard/standard/expressions.md#1147-numeric-promotions) section of the [C# language specification](~/_csharpstandard/standard/README.md). The `++` and `--` operators are defined for all integral and floating-point numeric types and the [char](../builtin-types/char.md) type. The `++`, `--`, and all compound arithmetic operators (such as `+=`), dont convert their arguments to `int` or have the result type as `int`.
```
